monix-eval/shared/src/main/scala/monix/eval/tracing/TaskTrace.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2014-2020 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.eval.tracing
+
+import monix.eval.Task
+
+import scala.reflect.NameTransformer
+
+/**
+  * All Credits to https://github.com/typelevel/cats-effect and https://github.com/RaasAhsan
+  */
+final case class TaskTrace(events: List[TaskEvent], captured: Int, omitted: Int) {
```

Comment:
I'm wondering whether classes like that should be reused with Cats-Effect, or not.
I suppose it's safer to have it separate and create a type class if there are any libraries that operate on it.

Reply:
Another idea would be to slurp this out into a common library which both can share. You would probably also be interested in the `ConcurrentHashMap` replacement that we're working on for 2.2.1, which should shrink the performance hit pretty dramatically.

Reply:
@djspiewak that looks interesting, where can I learn more about what replacement you're preparing for `ConcurrentHashMap`?

Reply:
@jvican https://github.com/typelevel/cats-effect/pull/1101

It's not a *general* replacement. CHM is actually insanely good at what it does. But tracing is a rather weird case. The cache converges to a relatively small size (strictly bounded by the number of classes) and then usually never changes again. Cache misses aren't fatal, they're just poorly optimized, and writes are not *mandatory* (follows from the miss bit). But within the context of all of that, we want reads to be *blisteringly* fast once fully converged.

In our synthetic benchmarks, tracing imposes a roughly 16% performance penalty on programs' straight-line performance (i.e. long chains of `flatMap`s). From what we can tell, nearly all of that is tied up in just *reading* from `ConcurrentHashMap`, due to the memory barriers which are necessary to ensure proper consistency, and secondarily due to the fact that the table doesn't aggressively optimize for a small and convergent set of keys. We can relax both of those constraints with a custom implementation and theoretically shave quite a bit off of that performance penalty.

Still a work in progress. :-)

Reply:
> Another idea would be to slurp this out into a common library which both can share.

That would be nice. Most of the code seems to be nicely separated and quite generic

Currently, I'm fine with copy-pasting and keeping an eye on updates but I will probably get tired of it pretty quickly, considering I plan to do the same for `Coeval` and `monix-bio` in the future

> You would probably also be interested in the ConcurrentHashMap replacement that we're working on for 2.2.1, which should shrink the performance hit pretty dramatically.

Sounds awesome!
Could you use any help?

Reply:
@djspiewak thanks for elaborating, will follow that thread. I asked because I, too, believe CHM is incredibly good at what it does and it's an amazing fit for many broad use cases. It makes sense to try to go for a barrier-free approach for tracing.

Reply:
> That would be nice. Most of the code seems to be nicely separated and quite generic
>
> Currently, I'm fine with copy-pasting and keeping an eye on updates but I will probably get tired of it pretty quickly, considering I plan to do the same for Coeval and monix-bio in the future

We should definitely pull it upstream into a new lib then. This is also relevant for CE3, where we want to reimplement tracing, but it would be nice to not have to duplicate things. Ideas on a name?

Reply:
👍  on factoring out code for reusability. I would start with something like the `UnsafeConvergentCache` because it's actually very general-purpose and initialization-safe, thread-unsafe data structures may have applications outside tracing. Pulling out tracing-specific code like `IOTrace`/`TaskTrace` is also an awesome idea, but my only caution is that functionality will most certainly evolve in the CE2/CE3/Monix world (e.g. thread tracking will be introduced to CE2 soon), and saying that they won't diverge is probably short-sighted. It would be unfortunate to end up in a situation where the library prescribes too much on implementations or implementations are restricted by the library, but I think we can definitely find a middle ground. Ultimately if things don't work out, we should be able to sever the dependency easily (keeping use of that library internal is probably a reasonable precaution).

> Currently, I'm fine with copy-pasting and keeping an eye on updates but I will probably get tired of it pretty quickly, considering I plan to do the same for Coeval and monix-bio in the future

If there are any improvements, it would be awesome to see those back in CE as well! :D But that's a really good reason to share all this code.

>  I rather like Sacagawea (tracking, tracing, etc), partially because Native American names and words are sorely underrepresented in the software space, and partially because she's cool and it seems like a distinctive name.

Sacagawea is an awesome name :)

Reply:
> Sacagawea is an awesome name :)

I can create a repo and set up some infra for it a little later this morning and you guys can shovel things in there as you see fit. :-) If it turns out to not be a good idea, it's always easy to un-shovel.

Reply:
https://github.com/typelevel/sacagawea

Reply:
Okay, I think I'll try to settle this PR as is, and then we could move towards keeping internals in `sacagawea`. Hopefully, I can find some improvements to pay you back a little. :)

I've counted only three public pieces (apart from configuration):

- `final case class IOTrace(events: List[IOEvent], captured: Int, omitted: Int)`
- `IOEvent`
- `PrintingOptions`

And it should be perfectly fine to keep it per library, mostly cosmetic stuff apart from `IOTrace`.
In case of `IOTrace`, we could just use a type class in any future libraries.

> Pulling out tracing-specific code like IOTrace/TaskTrace is also an awesome idea, but my only caution is that functionality will most certainly evolve in the CE2/CE3/Monix world (e.g. thread tracking will be introduced to CE2 soon)

Could you elaborate on this feature? 
I'm wondering about `IOTrace#showFiberTrace` for other types of events. 
Would it add extra lines about the current `Thread`?